snakemake/deployment/docker.py

```diff
@@ -0,0 +1,119 @@
+__author__ = "Johannes Köster"
+__copyright__ = "Copyright 2015-2020, Johannes Köster"
+__email__ = "johannes.koester@uni-due.de"
+__license__ = "MIT"
+
+import os
+import re
+import requests
+
+
+class DockerContainer:
+    """A Docker Container is a thin wrapper to a Docker Registry API to
+       primarily determine if a container exists to be used for a workflow
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.uri = None
+        if self.exists():
+            self.uri = "docker://%s" % self.name
+
+    @property
+    def version(self):
+        """Derive a container verison. Will by default be latest (the tag)
+           but can be a different tag or full hash.
+        """
+        # Does the container have a version (tag or hash?)
+        version = "latest"
+        if "@" in self.name:
+            container_name, version = self.name.split("@", 1)
+        elif ":" in self.name:
+            container_name, version = self.name.split(":", 1)
+        return version
+
+    @property
+    def registry(self):
+        """Return the container registry. We default to Docker Hub unless
+           another (quay.io) is defined
+        """
+        registry = "https://index.docker.io/v2"
```

Comment:
We might eventually add more registries that implement the Docker Distribution API - to start I thought Docker Hub and Quay.io were reasonable. But heads up that Docker Hub just announced they are going to start purging containers that aren't used (push or pull) after some time, so we definitely should prefer quay.